src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,452 @@
+ * for contiguous data over a computational domain, or a representation of
+ * fourier data.
+ *
+ * The VectorImpl does not itself support any particular mathematical
+ * operations on the contained values. The VectorImpl template class and the
+ * macros defined in VectorImpl.hpp assist in the construction of various
+ * derived classes supporting a chosen set of methematical operations.
+ *
+ */
+template <typename T, typename VECTOR_TYPE>
+class VectorImpl
+    : public PointerVector<T, blaze::unaligned, blaze::unpadded,
+                           blaze::defaultTransposeFlag, VECTOR_TYPE> {
+  /// \cond HIDDEN_SYMBOLS
+  static constexpr void private_asserts() noexcept {
+    static_assert(
+        std::is_nothrow_move_constructible<VectorImpl<T, VECTOR_TYPE>>::value,
+        "Missing move semantics");
+  }
+  /// \endcond
+ public:
+  using value_type = T;
+  using allocator_type = std::allocator<value_type>;
+  using size_type = size_t;
+  using difference_type = std::ptrdiff_t;
+  using BaseType = PointerVector<T, blaze::unaligned, blaze::unpadded,
+                                 blaze::defaultTransposeFlag, VECTOR_TYPE>;
+  static constexpr bool transpose_flag = blaze::defaultTransposeFlag;
+
+  using ElementType = T;
+  using TransposeType = VectorImpl<T, VECTOR_TYPE>;
+  using CompositeType = const VectorImpl<T, VECTOR_TYPE>&;
+
+  using BaseType::operator[];
+  using BaseType::begin;
+  using BaseType::cbegin;
+  using BaseType::cend;
+  using BaseType::data;
+  using BaseType::end;
+  using BaseType::size;
+
+  // @{
+  // Upcast to `BaseType`
+  const BaseType& operator~() const noexcept {
+    return static_cast<const BaseType&>(*this);
+  }
+  BaseType& operator~() noexcept { return static_cast<BaseType&>(*this); }
+    // @}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wshadow" /*ignore shadowing of size*/
+  /// Create with the given size and value.
+  ///
+  /// \param size number of values
+  /// \param value the value to initialize each element.
+  explicit VectorImpl(
+      size_t size, T value = std::numeric_limits<T>::signaling_NaN()) noexcept
+      : owned_data_(size, value) {
+    reset_pointer_vector();
+  }
+
+  /// Create a non-owning VectorImpl that points to `start`
+  VectorImpl(T* start, size_t size) noexcept
+      : BaseType(start, size), owned_data_(0), owning_(false) {}
+#pragma GCC diagnostic pop
+
+  /// Create from an initializer list of doubles. All elements in the
+  /// `std::initializer_list` must have decimal points
+  template <class C, Requires<cpp17::is_same_v<C, T>> = nullptr>
+  VectorImpl(std::initializer_list<C> list) noexcept
+      : owned_data_(std::move(list)) {
+    reset_pointer_vector();
+  }
+
+  /// Empty VectorImpl
+  VectorImpl() noexcept = default;
+  /// \cond HIDDEN_SYMBOLS
+  ~VectorImpl() = default;
+
+  /// \cond HIDDEN_SYMBOLS
+
+  // clang-tidy: calling a base constructor other than the copy constructor.
```

Comment:
I think we should move the definitions of the next two constructors, the assignment operators, and the pup function out of the body of the class. The reason is that any function defined in the body of the class is implicitly marked as inline, which you don't want for long functions.